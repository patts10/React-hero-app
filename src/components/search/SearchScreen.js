import { useMemo } from "react";
import { useLocation, useNavigate } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import { getHeroesByName } from "../../selectors/getHeroByName";
import { HeroCard } from "../hero/HeroCard";
import queryString from 'query-string'

export const SearchScreen = () => {

  const navigate = useNavigate();
  const location = useLocation();

  const { q = '' } = queryString.parse(location.search);

  const [values, handleInputChange, reset] = useForm({
    searchText: q,
  });

  const { searchText } = values;

  const heroesFiltered = useMemo( () => getHeroesByName(q), [ q ] );

  const handleSearch = (e) => {
    e.preventDefault();
    navigate(`?q=${ searchText }`);
  };

  return (
    <>
      <h1 className="my-5">Searches</h1>

      <div className="flex flex-row">

        <div className="w-4/12">
          <h1>Search</h1>
          <hr />

          <form onSubmit={handleSearch}>
            <input
              type="text"
              placeholder="Search hero"
              name="searchText"
              className="block px-2 py-1 mt-5 w-full"
              onChange={handleInputChange}
              value={searchText}
              autoComplete="off"
            />

            <button
              className="mt-3 px-6 py-2.5 bg-white text-blue-600 hover:text-white font-medium text-xs border border-blue-600 leading-tight rounded shadow-md hover:bg-blue-600 hover:shadow-lg focus:shadow-lg focus:outline-none focus:ring-0 active:bg-blue-600 active:shadow-lg transition duration-150 ease-in-out"
              type="submit"
            >
              Search
            </button>
          </form>
        </div>

        <div className="w-8/12">
          <h4>Results</h4>
          <hr />

          {
            (q==='')
              ? <div className="bg-blue-100 rounded-lg m-3 py-3 px-6 mb-4 text-base text-blue-700 mb-3" role="alert">Search a hero</div>
              : (heroesFiltered.length == 0)
                && <div className="bg-red-100 rounded-lg m-3 py-3 px-6 mb-4 text-base text-red-700 mb-3" role="alert">There is no results: { q }</div>
          }

          {
            heroesFiltered.map( hero => (
              <HeroCard
                key={ hero.id }
                { ...hero }
              />
            ))
          }

        </div>
      </div>
    </>
  );
};
