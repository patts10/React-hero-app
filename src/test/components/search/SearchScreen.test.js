import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import { SearchScreen } from "../../../components/search/SearchScreen";

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () =>({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate
}))

describe("Tests on <SearchScreen />", () => {
  test("should show default values correctly", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(".rounded-lg").text().trim()).toBe("Search a hero");
  });

  test("should show Batman and input value to be equal queryString", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("input").prop("value")).toBe("batman");
  });

  test("should show error if here is not found", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=batman123"]}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect(wrapper.find(".alert-danger").exists()).toBe(true);
    expect(wrapper.find(".alert-danger").text()).toBe(
      "There is no results: batman123"
    );
  });

  test("should call navigate to new screen", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <SearchScreen />
      </MemoryRouter>
    );

    wrapper.find('input').simulate('change', {
      target: {
        name: "searchText",
        value: "batman",
      }
    });

    wrapper.find('form').prop('onSubmit')({
      preventDefault: () => {}
    });

    expect( mockNavigate ).toHaveBeenCalled();
    expect( mockNavigate ).toHaveBeenCalledWith('?q=batman');
  });
});
