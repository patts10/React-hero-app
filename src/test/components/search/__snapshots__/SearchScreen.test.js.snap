// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <SearchScreen /> should show Batman and input value to be equal queryString 1`] = `
Array [
  <h1
    className="my-5"
  >
    Searches
  </h1>,
  <div
    className="flex flex-row"
  >
    <div
      className="w-4/12"
    >
      <h1>
        Search
      </h1>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="block px-2 py-1 mt-5 w-full"
          name="searchText"
          onChange={[Function]}
          placeholder="Search hero"
          type="text"
          value="batman"
        />
        <button
          className="mt-3 px-6 py-2.5 bg-white text-blue-600 hover:text-white font-medium text-xs border border-blue-600 leading-tight rounded shadow-md hover:bg-blue-600 hover:shadow-lg focus:shadow-lg focus:outline-none focus:ring-0 active:bg-blue-600 active:shadow-lg transition duration-150 ease-in-out"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="w-8/12"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="rounded-lg border border-gray-300 mb-4 text-center m-3 lg:w-3/12 md:w-2/5 animate__animated animate__fadeIn"
      >
        <div
          className="flex flex-row"
        >
          <div
            className="w-4/12"
          >
            <img
              alt="Batman"
              src="assets/dc-batman.jpg"
            />
          </div>
          <div
            className="flex items-center w-8/12"
          >
            <div
              className="py-4 w-full text-left px-3"
            >
              <h5
                className="md:text-2xl sm:text-2xl"
              >
                Batman
              </h5>
              <p
                className="md:text-base sm:text-sm mt-3"
              >
                Bruce Wayne
              </p>
              <p>
                <small
                  className="text-gray-600"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                className="hover:underline hover:text-blue-900"
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                more...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Tests on <SearchScreen /> should show default values correctly 1`] = `
Array [
  <h1
    className="my-5"
  >
    Searches
  </h1>,
  <div
    className="flex flex-row"
  >
    <div
      className="w-4/12"
    >
      <h1>
        Search
      </h1>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="block px-2 py-1 mt-5 w-full"
          name="searchText"
          onChange={[Function]}
          placeholder="Search hero"
          type="text"
          value=""
        />
        <button
          className="mt-3 px-6 py-2.5 bg-white text-blue-600 hover:text-white font-medium text-xs border border-blue-600 leading-tight rounded shadow-md hover:bg-blue-600 hover:shadow-lg focus:shadow-lg focus:outline-none focus:ring-0 active:bg-blue-600 active:shadow-lg transition duration-150 ease-in-out"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="w-8/12"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="bg-blue-100 rounded-lg m-3 py-3 px-6 mb-4 text-base text-blue-700 mb-3"
        role="alert"
      >
        Search a hero
      </div>
    </div>
  </div>,
]
`;
